### ISSUE TITLE
Redesigning

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/testing#8]
I tried to avoid it, but #7 convinced me to redesign the library. The biggest change is having a bigger Test class that covers multiple async functions containing tests. So if we have an example like this:

```js
{
    async testA(){},
    async testB(){},
    async testC(){}
}
```

then we will end up with this structure:

```js
new Test({
    async testA(){},
    async testB(){},
    async testC(){}
});
```

instead of

```js
new Group({
    testA: new Test(async function testA(){}),
    testB: new Test(async function testB(){}),
    testC: new Test(async function testC(){})
});
```

**Comment by inf3rno** [comment]
From one perspective this appears to be a lot better, because it is easier to say to use the Test class in the test files instead of just exporting an object. This would make auto completion a lot easier too. On the other hand it makes things like skipping tests somewhat harder and adds some group related code to the tests too.

**Comment by inf3rno** [comment]
Probably we started with the wrong approach by saying that a Test contains a very simple test in a single execute method and we should group our tests using the Group class just to be able to run and find them easier and to share variables in a cumbersome way.

Maybe a Test should be a group instead that contains these small tests and makes sharing variables straightforward through class properties. We can still use the execute method to run our tests for example in a specific order. So not the executor or the upper group decides about test order at least not in such a test aggregate.

**Comment by inf3rno** [comment]
One problematic thing is that we need the same logic for skipping, ordering, etc. our tests and groups. This means for example when we have a dependency, then we need to run the dependent test only after its dependencies are done. The same should be true for test groups I think, but a test should not know about other tests and a group should not know about other groups in theory. In practice this is not true. We don't say from the otherside that ES6 depends on ES5, but we say in the ES6 tests that they should run only after we are done with the ES5 tests. Thought this means that we should load the ES6 tests even if we know we won't run their ES5 dependency and we will skip them. If we would define the dependency in the upper group we would never load them. I think this is a compromise I can accept if we have a much better code in exchange. I thought about something like

```js
{
    async a(){
        ...
    },
    async b(){
        await this.a();
    }
}
```

or in the case of groups:

```js
{
    async execute(){
        await this.t1.execute();
        this.t2.execute();
    }
}
```

**Comment by inf3rno** [comment]
Another question was how to ensure that the dependencies don't run twice?

For example:
```js
{
    async a(){
        ...
    },
    async b(){
        await this.a();
    },
    async execute(){
        Promise.all([this.a(), this.b()])
    }
}
```
Here "a" runs twice. One solution to keep the simple parallel code is returning the same promise for each call

```js
    async a(){
        if (!this.aResult)
            this.aResult = new Promise(...);
        return this.aResult;
    }
```

Thought I am not sure if we await an already resolved promise we will really get the result or it will just wait indefinitely. I need to check this. 99% that we will get the result, because `Promise.resolve()` does exactly this and it works. But I need to be certain.

Another way to handle this is not keeping the parallel execution in the execute method.

```js
    async execute(){
        await this.a();
        await this.b();
    }
```

This is not necessarily the best solution or we need a general code to handle scenarios where there are many dependencies. Otherwise we can end up blocking tests while waiting for dependencies of a different test.

For example here we can say that there are "c" and "d" methods too and "d" depdends on "c".

We can come up with completely wrong solutions:

```js
    async execute(){
        await this.a();
        await this.c();
        await this.b();
        await this.d();
    }
```

```js
    async execute(){
        await Promise.all([this.a(), this.c()]);
        await Promise.all([this.b(), this.d()]);
    }
```

We'd need something like

```js
    async execute(){
        await Promise.all(
            Promise.series([this.a(), this.b()]),
            Promise.series([this.c(), this.d()])
        );
    }
```

But we don't have the `Promise.series` method defined, so we need to build such a thing. Another problem here if we want to pass the results. We would end up with an async library like this: https://github.com/caolan/async which I certainly don't want to do.

```js
    async execute(){
        await Promise.all(
            (async () => {
                await this.a();
                await this.b();
            })(),
            (async () => {
                await this.c();
                await this.d();
            })(),
        );
    }
```

Something like this would be the right solution with current js tools. I find it pretty ugly compared to leaving the execute alone and calling the dependency from the dependent test.

**Comment by inf3rno** [comment]
While this works pretty well in the case of these test classes. We have a problem in the case of groups. We need to refer somehow the group we depend on in the execute, but that cannot be done with a simple `await this.x()` call because only the upper level group knows about our dependency. We need to inject our dependency or ask the upper level group to await it before our group runs and we need to inject the upper level group or some sort of helper object / function which helps with it.

**Comment by inf3rno** [comment]
A simple solution for tests looks like this:

```js
export default class Test {
    constructor(config){
        if (config)
            Object.assign(this, config);
    }
    async execute(){
        var tests = [...this.getTestNames()].map((key) => this[key]);
        await Promise.all(tests);
        return "ok";
    }
    getTestNames(){
        var reservedNames = Object.getOwnPropertyNames(Test.prototype);
        var testNames = new Set();
        walkPrototypeChain(this, function (object){
            if (object === Object.prototype || object === Test.prototype)
                return;
            for (var property of Object.getOwnPropertyNames(object))
                if (reservedNames.indexOf(property) == -1 && (this[property] instanceof Function) && util.types.isAsyncFunction(this[property]))
                    testNames.add(property);
        }.bind(this));
        return testNames;
    }
}
```

**Comment by inf3rno** [comment]
Here we can decide if we want to support dependencies somehow. For example 
```js
async execute(...parameters){
    this.addDependency(this.b, this.a);
    return super.execute(...parameters);
}
```

**Comment by inf3rno** [comment]
While I like the 

```js
    async execute(){
        await Promise.all(
            (async () => {
                await this.a();
                await this.b();
            })(),
            (async () => {
                await this.c();
                await this.d();
            })(),
        );
    }
```

approach, I think it is better to have a general execute and have dependencies. I wanted the execution to be customizable, but the more I think about it the better it looks like to have a general solution for the common problems.

Another solution to add helpers to the upper approach
```js
    async execute(){
        return this.parallel(
            this.series(this.a, this.b),
            this.series(this.c, this.d)
        );
    }
```

The main problem with this, that we don't have a list of tests, so we need to add all by name one by one. I think we should have some sort of test pool.

```js
    async execute(){
        return this.parallel(
            this.series(this.a, this.b),
            this.series(this.c, this.d),
            this.rest()
        );
    }
```

or

```js
    async execute(){
        return this.parallel(
            this.series(this.a, this.b),
            this.series(this.c, this.d),
            this.all().remove(this.a, this.b, this.c, this.d)
        );
    }
```

I find these ugly compared to the dependency management.

```js
    async execute(){
        this.addDependency(this.b, this.a);
        this.addDependency(this.d, this.c);
        return super.execute();
    }
```

At least it is on a different abstraction level. The dependency is more abstract than the execution order.

**Comment by inf3rno** [comment]
What I would like better is the following:

```js
    async execute(...args){
        this.executor.select(this.b).dependsOn(this.a);
        this.executor.select(this.d).dependsOn(this.c);
        return this.executor.run(...args);
    }
```

**Comment by inf3rno** [comment]
We can have a GroupExecutor and a TestExecutor, but we can have a single Executor class too by adding and if-else for async functions and groups. So here always the upper level decides about the dependency, not the lower level. Is there a way to support somehow the latter?

**Comment by inf3rno** [comment]
```js
{
    async a(){
        if (!this.aResult)
            this.aResult = new Promise(...);
        return this.aResult;
    },
    async b(){
        await this.a();
    },
    async execute(){
        return Promise.all([this.a(), this.b()]);
    }
}
```

Here all we can do is wrapping "a" with something like `.once`. Actually we should do the same with all our tests I think. Not sure exactly how async functions work regarding to return values when I add an await in the middle. If they have a problem with it, then we still can do Promise.exposed with my lib.

**Comment by inf3rno** [comment]
How to execute these tests twice?

We could move the whole wrapping to an executor, but if the test functions keep state in the test instance, then that can cause problems when we run them multiple times. Another issue here is using the helper functions. We need to copy those somehow too. I think the best would be using a Proxy and copying the test instance with Object.create if we don't want to run it on the original one.

**Comment by inf3rno** [comment]
One thing that makes the `await this.a()` approach problematic that when "a" fails, "b" would fail too. Normally we would just skip "b" and add something like failing dependency, but we cannot do that here. Well actually we can by wrapping an rethrowing the failure. Something like `TestFailed("a", err)` is fine, because then we would throw `TestFailed("b", TestFailed("a", err))` from "b", and that would mean a failing dependency. So this can be solved by wrappers without configuration.

**Comment by inf3rno** [comment]
One question remains, how to do group and/or test instance dependency?

Well we can inject the upper level group, which is ugly, or we can inject the dependency.

```js
class es6 {
    constructor(es5){
        this.es5 = es5;
        this.wrapTests();
    }
    async execute(...params){
        await this.es5();
        return Promise.all(this.tests.map((test)=>test.apply(...params)));
    }
}
```

**Comment by inf3rno** [comment]
So the test group would look like

```js
˙class ES extends Group {
    constructor(){
        this.es5 = new ES5();
        this.es6 = new ES6(this.es5);
    }
}
```

We can go even further by making the dependencies more detailed:

```js
class B {
    constructor(a){
        this.a= a;
        this.wrapTests();
    }
    async y(){
        await this.a.x();
        ...
    }
}
```

In theory we can prevent running the whole dependency group and we could rely on the individual tests. If we need the whole group, then we should do

```js
class B {
    constructor(a){
        this.a= a;
        this.wrapTests();
    }
    async y(){
        await this.a.execute();
        ...
    }
}
```

or

```js
class B {
    constructor(a){
        this.a= a;
        this.wrapTests();
    }
    async execute(){
        await this.a.execute();
        ...
    }
}
```

**Comment by inf3rno** [comment]
So with dependency injection we can avoid configuring group dependencies. This is good, because usually we know by writing a test whether it depends on another one and we don't want to go to the upper level code if not necessary. We can use a fully automated DI container to inject dependencies by parameter name.

**Comment by inf3rno** [comment]
As of the wrapping and the automatic test discovery I think it should be customizable, which means we'd better to move those to separate methods, something like `isTest(value, property, prototype)` and `runsOnce(value, property, prototype)`. We need to wrap them anyways, because we need to add a try-catch block for error handling. By default `runsOnce` should return true for all tests and we can add exceptions in the form of `if (value === this.a) return false;`.

**Comment by inf3rno** [comment]
We might want to add a proxy for dependencies, so we could collect some info in the wrapper about dependencies, which can be useful for our reporter.

**Comment by inf3rno** [comment]
Now we just need to decide about how to skip tests. The main problem is dependencies here. Since we know about dependencies partially on the fly by calling wrappers, we need to add the code for skip to those. So it is not an option to avoid loading dependencies, but we can prevent running them. We can have 2 possible way to handle skipped depenencies. One way is running them anyways, because the actual test needs them. The other way is skipping them and hope that the test does not fail because a dependency is not met. The second solution is necessary, when we already ran the dependency so we are pretty sure that it is fine, but running it every time is slow. So in reality what we really need is a cache rather than skipping it. This means we can solve this dependency skipping by running the dependencies anyways but having a cache for the second run. We can do this by using a test server for running, so we can keep the cache in memory, or we can save it into a cache file, preferably JSON or my appendable JSON format. In this case the reporter should get a cached flag for the results. Another solution is skipping both the dependency and the dependent. I think we need to make the skipping system very flexible to support both.

For now I think I'll implement the library without skipping and add it only later. As of the wrappers I think I'll use a proxy as context. That way I don't have to override the actual methods and it is easier to add general wrapper code.

**Comment by inf3rno** [comment]
I still need to decide to support `new Test(...)` or just `class My extends Test {...}`. I thought about making the container manually or in a semi-manual way. That would solve skipping for example. It would be nice to do at least once manually and check if we can automate it, so I need a proof of concept code before doing the actual abstractions. Otherwise I cannot properly write the wiki. Well I thnk I'll still write the wiki first and change it accordingly when I run into a coding problem.

**Comment by inf3rno** [comment]
Well we don't necessarily need injection. If all we need is a simple hierarchy, then we can just instantiate the other tests in our constructor. If we have a directed acyclic dependency graph instead of a dependency tree, then we need to inject. I am not sure if automation worth the effort. I mean it really depends on how much time it is to maintain our graph and add some extra classes to do it. In many cases we don't have enough tests to have a real issue and we can do it manually.

**Comment by inf3rno** [comment]
So I think our 0.1 version should support only manual dependency handling via injection and direct calls from the dependent tests. Our reporter should get the same events as it gets with our previous model. There will be differences because we won't have a plan I think, or not in the form as our previous model. It would be more like something that contains what to skip and what to run, so it will be something like the runner in our previous model, if we had that. I don't remember.

**Comment by inf3rno** [comment]
Okay, the conclusion here, that we will give the dependencies in the code instead of using a config solution.

**Comment by inf3rno** [comment]
I'll open separate issues for the other questions.
 - do we extend our test class or instantiate it when giving our tests?
 - should we add a once wrapper by default or that should be the exception?
 - what  rules do we have by skipping dependencies?
 - do we use a DI container for injecting dependencies?
 - do we allow instantiation of sub-tests inside the test or do we use a container for everything?
 - do we use the file system hierarchy or the dependency graph or something else for reporting?
 - how do we find a test that we want to skip? should we use tagging?

**Comment by inf3rno** [comment]
I wanted real life examples, so I copied a test from one of my projects:

```js
var expect = require("expect.js"),
    sinon = require("sinon"),
    polyfill = require("../.");

describe('Capturing the stack trace', function () {

    describe('writing the current stack trace on throwables', function () {
        context('when I write the stack trace onto a throwable', function () {
            var aThrowable = {
                name: "Throwable",
                message: "some text"
            };
            var firstFn = function firstFn() {
                secondFn();
            };
            var secondFn = function secondFn() {
                thirdFn();
            };
            var thirdFn = function thirdFn() {
                Error.captureStackTrace(aThrowable);
            };
            firstFn();

            it('should store information about the present stack', function (next) {
                expect(Error.getStackTrace(aThrowable)).to.not.be(undefined);
                next();
            });
            it('should return the formatted stack by reading this information', function (next) {
                var stackString = Error.getStackTrace(aThrowable);
                expect(typeof(stackString)).to.be("string");
                expect(/Throwable: some text/.test(stackString)).to.be.ok();
                var frameStrings = stackString.split("\n");
                frameStrings.shift();
                expect(/thirdFn/.test(frameStrings[0])).to.be.ok();
                expect(/secondFn/.test(frameStrings[1])).to.be.ok();
                expect(/firstFn/.test(frameStrings[2])).to.be.ok();
                next();
            });
        });
    });
});
```

**Comment by inf3rno** [comment]
I can simplify this with the following code:

```js
var expect = require("expect.js"),
    sinon = require("sinon"),
    polyfill = require("../.");

{
	'Capturing the stack trace': {
		'writing the current stack trace on throwables': {
			'when I write the stack trace onto a throwable': {
				setup(){
					var aThrowable = {
						name: "Throwable",
						message: "some text"
					};
					var firstFn = function firstFn() {
						secondFn();
					};
					var secondFn = function secondFn() {
						thirdFn();
					};
					var thirdFn = function thirdFn() {
						Error.captureStackTrace(aThrowable);
					};
					firstFn();
					this.stackString = Error.getStackTrace(aThrowable);
				},
				'should store information about the present stack': function (next) {
					expect(this.stackString).to.not.be(undefined);
					next();
				},
				'should return the formatted stack by reading this information': function (next) {
					expect(typeof(this.stackString)).to.be("string");
					expect(/Throwable: some text/.test(this.stackString)).to.be.ok();
					var frameStrings = this.stackString.split("\n");
					frameStrings.shift();
					expect(/thirdFn/.test(frameStrings[0])).to.be.ok();
					expect(/secondFn/.test(frameStrings[1])).to.be.ok();
					expect(/firstFn/.test(frameStrings[2])).to.be.ok();
					next();
				}
			}
		}
    }
}
```

**Comment by inf3rno** [comment]
I don't like this because of too many indentations. A fluent interface would be so much better.
